src/Security/Authentication/Negotiate/src/NegotiateHandler.cs

```diff
@@ -0,0 +1,334 @@
+                    if (_negotiateState?.IsCompleted == false)
+                    {
+                        throw new InvalidOperationException("Non-negotiate request was received in between authentication handshake requests.");
+                    }
+                    Logger.LogDebug($"Non-Negotiate Authorization header");
+                    return false;
+                }
+
+                // WinHttpHandler re-authenticates an existing connection if it gets another challenge on subsequent requests.
+                if (_negotiateState?.IsCompleted == true)
+                {
+                    Logger.LogDebug("Negotiate data received for an already authenticated connection, Re-authenticating");
+                    _negotiateState.Dispose();
+                    _negotiateState = null;
+                    persistence.State = null;
+                }
+
+                _negotiateState ??= Options.StateFactory.CreateInstance();
+
+                var outgoing = _negotiateState.GetOutgoingBlob(token);
+
+                if (!_negotiateState.IsCompleted)
+                {
+                    if (_negotiateState.Protocol == "NTLM")
+                    {
+                        persistence ??= EstablishConnectionPersistence(connectionItems);
+                        // Save the state long enough to complete the multi-stage handshake.
+                        // We'll remove it once complete if !PersistNtlmCredentials.
+                        persistence.State = _negotiateState;
+                    }
+
+                    Logger.LogInformation("Incomplete Negotiate, 401 Negotiate challenge");
+                    Response.StatusCode = StatusCodes.Status401Unauthorized;
+                    Response.Headers.Append(HeaderNames.WWWAuthenticate, AuthHeaderPrefix + outgoing);
+                    return true;
+                }
+
+                if (string.IsNullOrEmpty(outgoing))
+                {
+                    Logger.LogDebug($"Completed Negotiate.");
+                }
+                else
+                {
+                    Logger.LogDebug($"Completed Negotiate with additional data.");
+                    // There can be a final blob of data we need to send to the client, but let the request execute as normal.
+                    Response.Headers.Append(HeaderNames.WWWAuthenticate, AuthHeaderPrefix + outgoing);
+                }
+
+                // Deal with connection credential persistence.
+
+                if (_negotiateState.Protocol == "NTLM" && !Options.PersistNtlmCredentials)
+                {
+                    // NTLM was already put in the persitence cache on the prior request so we could complete the handshake.
+                    // Take it out if we don't want it to persist.
+                    Debug.Assert(object.ReferenceEquals(persistence?.State, _negotiateState),
+                        "NTLM is a two stage process, it must have already been in the cache for the handshake to succeed.");
```

Comment:
You are using an IOptionsMonitor which means the persistance can change. Debug.Assert is meant for when there's a problem with unexpected state indicating a bug but you can get here simply by having the option value change.

Reply:
Options changing under us is an interesting topic, but it doesn't affect this assert.

This assert pairs with line 140 above saying all NTLM handshakes require multiple requests and the incomplete state was temporarily persisted regardless of the value of PersistNtlmCredentials.

If the persistence option was changed between requests then you could have two things happen:
- Persistence enabled: No problem, nothing was persisted before, the next request will persist.
- Persistence disabled: Existing connections will maintain their persistence, new connections will not. I don't see anything here that will break.